File: src/main/java/com/bajdcc/LALR1/grammar/tree/ExpAssign.java
package com.bajdcc.LALR1.grammar.tree;

import com.bajdcc.LALR1.grammar.codegen.ICodegen;
import com.bajdcc.LALR1.grammar.runtime.RuntimeInst;
import com.bajdcc.LALR1.grammar.semantic.ISemanticRecorder;
import com.bajdcc.LALR1.grammar.tree.closure.IClosureScope;
import com.bajdcc.util.lexer.token.KeywordType;
import com.bajdcc.util.lexer.token.OperatorType;
import com.bajdcc.util.lexer.token.Token;

/**
 * 【语义分析】赋值表达式
 *
 * @author bajdcc
 */
public class ExpAssign implements IExp {

	/**
	 * 变量名
	 */
	private Token name = null;

	/**
	 * 表达式
	 */
	private IExp exp = null;

	/**
	 * 是否为声明
	 */
	private boolean decleared = false;

	public Token getName() {
		return name;
	}

	public void setName(Token name) {
		this.name = name;
	}

	public IExp getExp() {
		return exp;
	}

	public void setExp(IExp exp) {
		this.exp = exp;
	}

	public boolean isDecleared() {
		return decleared;
	}

	public void setDecleared(boolean decleared) {
		this.decleared = decleared;
	}

	@Override
	public boolean isConstant() {
		return false;
	}

	@Override
	public boolean isEnumerable() {
		return false;
	}

	@Override
	public IExp simplify(ISemanticRecorder recorder) {
		return this;
	}

	@Override
	public void analysis(ISemanticRecorder recorder) {
		if (exp != null)
			exp.analysis(recorder);
	}

	@Override
	public void genCode(ICodegen codegen) {
		if (exp != null)
			exp.genCode(codegen);
		else
			codegen.genCode(RuntimeInst.ipushx);
		codegen.genCode(RuntimeInst.ipush, codegen.genDataRef(name.object));
		if (decleared) {
			codegen.genCode(RuntimeInst.ialloc);
		} else {
			codegen.genCode(RuntimeInst.istore);
		}
	}

	@Override
	public String toString() {
		return print(new StringBuilder());
	}

	@Override
	public String print(StringBuilder prefix) {
		return (decleared ? KeywordType.VARIABLE.getName() : KeywordType.LET
				.getName()) +
				" " + name.toRealString() +
				(exp != null ? (" " + OperatorType.ASSIGN.getName() + " " +
				exp.print(prefix)) : "");
	}

	@Override
	public void addClosure(IClosureScope scope) {
		if (decleared) {
			scope.addDecl(name.object);
		}
		if (exp != null)
			exp.addClosure(scope);
	}

	@Override
	public void setYield() {

	}
}

File: src/main/kotlin/com/bajdcc/LALR1/grammar/tree/ExpAssign.kt
package com.bajdcc.LALR1.grammar.tree

import com.bajdcc.LALR1.grammar.codegen.ICodegen
import com.bajdcc.LALR1.grammar.runtime.RuntimeInst
import com.bajdcc.LALR1.grammar.semantic.ISemanticRecorder
import com.bajdcc.LALR1.grammar.tree.closure.IClosureScope
import com.bajdcc.util.lexer.token.KeywordType
import com.bajdcc.util.lexer.token.OperatorType
import com.bajdcc.util.lexer.token.Token

/**
 * 【语义分析】赋值表达式
 *
 * @author bajdcc
 */
class ExpAssign : IExp {

    /**
     * 变量名
     */
    var name: Token? = null

    /**
     * 表达式
     */
    var exp: IExp? = null

    /**
     * 是否为声明
     */
    var isDecleared = false

    override fun isConstant(): Boolean {
        return false
    }

    override fun isEnumerable(): Boolean {
        return false
    }

    override fun simplify(recorder: ISemanticRecorder): IExp {
        return this
    }

    override fun analysis(recorder: ISemanticRecorder) {
        if (exp != null)
            exp!!.analysis(recorder)
    }

    override fun genCode(codegen: ICodegen) {
        if (exp != null)
            exp!!.genCode(codegen)
        else
            codegen.genCode(RuntimeInst.ipushx)
        codegen.genCode(RuntimeInst.ipush, codegen.genDataRef(name!!.obj!!))
        if (isDecleared) {
            codegen.genCode(RuntimeInst.ialloc)
        } else {
            codegen.genCode(RuntimeInst.istore)
        }
    }

    override fun toString(): String {
        return print(StringBuilder())
    }

    override fun print(prefix: StringBuilder): String {
        return (if (isDecleared)
            KeywordType.VARIABLE.desc
        else
            KeywordType.LET
                    .desc) +
                " " + name!!.toRealString() +
                if (exp != null)
                    " " + OperatorType.ASSIGN.desc + " " +
                            exp!!.print(prefix)
                else
                    ""
    }

    override fun addClosure(scope: IClosureScope) {
        if (isDecleared) {
            scope.addDecl(name!!.obj!!)
        }
        if (exp != null)
            exp!!.addClosure(scope)
    }

    override fun setYield() {

    }
}